frontera/contrib/backends/remote/codecs/json.py

```diff
@@ -3,29 +3,59 @@
 """
 from __future__ import absolute_import
 import json
+import six
 from base64 import b64decode, b64encode
 from frontera.core.codec import BaseDecoder, BaseEncoder
-from w3lib.util import to_unicode, to_native_str
-from frontera.utils.misc import dict_to_unicode, dict_to_bytes
+from w3lib.util import to_unicode, to_bytes
+
+
+def _encode_recursively(obj):
+    """
+    recursively encodes an object to unicode and store its type
+    """
+    if isinstance(obj, bytes):
+        return 'bytes', to_unicode(obj)
+    elif isinstance(obj, dict):
+        return 'dict', [(_encode_recursively(k), _encode_recursively(v)) for k, v in six.iteritems(obj)]
+    elif isinstance(obj, (list, tuple)):
+        return type(obj).__name__, [_encode_recursively(item) for item in obj]
+    return 'other', obj
+
```

Comment:
A dict can can have nested dicts or other  kinds of objects. Since we are doing some kind of conversion here, we want to save the type here for successful de-conversion.  Json Encoder taking care is not sufficient. We need to successfully Decode this to the original form for which we need to save type.
Although this implementation currently only converts strings, list and dicts it can be easily extended for something like sets. 
If you have any other implementation in mind, please let me know.